### ISSUE TITLE
Isolate bazel_worker_java dependencies

### ISSUE THREAD

**Issue by Bencodes** [bazelbuild/bazel-worker-api#7]
We see lots of warnings in our console builds because of these rules relying on the default rules_jvm_external `maven` namespaces.

```console
(15:01:37) DEBUG: /private/var/tmp/_bazel_blee/499a001013731d09bffd82f8601a3161/external/rules_jvm_external+/private/extensions/maven.bzl:155:14: The maven repository 'maven' is used in two different bazel modules, originally in 'protobuf' and now in 'bazel_worker_java'
(15:01:37) DEBUG: /private/var/tmp/_bazel_blee/499a001013731d09bffd82f8601a3161/external/rules_jvm_external+/private/extensions/maven.bzl:155:14: The maven repository 'maven' is used in two different bazel modules, originally in 'protobuf' and now in 'bazel_worker_java'
(15:01:37) DEBUG: /private/var/tmp/_bazel_blee/499a001013731d09bffd82f8601a3161/external/rules_jvm_external+/private/extensions/maven.bzl:155:14: The maven repository 'maven' is used in two different bazel modules, originally in 'protobuf' and now in 'bazel_worker_java'
(15:01:37) DEBUG: /private/var/tmp/_bazel_blee/499a001013731d09bffd82f8601a3161/external/rules_jvm_external+/private/extensions/maven.bzl:155:14: The maven repository 'maven' is used in two different bazel modules, originally in 'protobuf' and now in 'bazel_worker_java'
(15:01:37) DEBUG: /private/var/tmp/_bazel_blee/499a001013731d09bffd82f8601a3161/external/rules_jvm_external+/private/extensions/maven.bzl:155:14: The maven repository 'maven' is used in two different bazel modules, originally in 'protobuf' and now in 'bazel_worker_java'
(15:01:37) DEBUG: /private/var/tmp/_bazel_blee/499a001013731d09bffd82f8601a3161/external/rules_jvm_external+/private/extensions/maven.bzl:155:14: The maven repository 'maven' is used in two different bazel modules, originally in 'protobuf' and now in 'bazel_worker_java'
````

**Comment by meteorcloudy** [comment]
We could use a different maven namespace for this module, but wouldn't there be risks of conflict classes? If your target A depends on bazel-worker-api, which depends on guava, and your target also directly depends on a different version of guava from another maven namespace.

@jin @shs96c What is the recommended best practice?

**Comment by jin** [comment]
Yes, care has to be taken when using multiple namespaces (https://github.com/bazel-contrib/rules_jvm_external?tab=readme-ov-file#multiple-maven_install-declarations-for-isolated-artifact-version-trees) that the dep classes from the different namespaces don't make it into the same top level target, i.e. enforcing the one-version rule at the target level.

Is there warning too verbose? We can consider improving it if that's the main concern.

**Comment by Bencodes** [comment]
> Is there warning too verbose? We can consider improving it if that's the main concern.

The messages are definitely verbose. Any time you run a Bazel command where it has to bring up the server for the first time, the console gets hit with a flood of warnings from `rules_jvm_external` all caused by this worker using the default namespace. I only pasted a subset of them above as an example.

**Comment by shs96c** [comment]
This can be fixed by updating to v29.1 of `protobuf`. Relevant issues are https://github.com/protocolbuffers/protobuf/pull/19477#event-15498868869 and https://github.com/protocolbuffers/protobuf/issues/19451#issuecomment-2520924027

**Comment by meteorcloudy** [comment]
The problem also exists in this repo:https://github.com/bazelbuild/bazel-worker-api/blob/3dfa3826c7e5706391aa1af9382c1ed3507e154b/java/MODULE.bazel#L39

Should we also rename it?

**Comment by shs96c** [comment]
I would suggest doing so if the repos are independent of each other.

**Comment by meteorcloudy** [comment]
Say if I have a target:
```
java_binary(
  name = "foo",
  srcs = ["Foo.java"],
  deps = [
     "@bazel_worker_java//src/main/java/com/google/devtools/build/lib
/worker:work_request_handlers",
     "@maven//:com_google_guava_guava",
  ],
)
```

If we rename the maven install for `bazel_woker_java`, wouldn't two different guava versions ended up for foo? Is that in general fine with Java?

**Comment by jin** [comment]
I don't think renaming is the right solution, because https://github.com/bazelbuild/bazel-worker-api/blob/3dfa3826c7e5706391aa1af9382c1ed3507e154b/java/src/main/java/com/google/devtools/build/lib/worker/BUILD.bazel#L6 is meant to be depended upon by users' java tooling targets, which means `:work_request_handlers`'s deps should also be contributed into default user `@maven` namespace.

**Comment by meteorcloudy** [comment]
@jin Thanks for the clarification!

**Comment by jin** [comment]
https://github.com/bazel-contrib/rules_jvm_external/pull/1295 will fix the noisy logs.

**Comment by Bencodes** [comment]
An alternative solution here would be to isolate the worker maven dependencies into it's own known maven repository, and then document how consumers of these rules can override the actual versions to match the versions that they define in their projects.

Example:
```
maven.override(
    name = "worker_maven",
    coordinates = "com.google.protobuf:protobuf-java",
    target = "@maven//:com_google_protobuf_protobuf_java",
)
```

**Comment by shs96c** [comment]
Finally getting around to properly looking at this issue. I don't think it's a good idea for every user of `rules_jvm_external` to pull in whatever it is that `bazel_worker_java` depends on into their dependencies. This will cause every user to repin their deps when they update (which isn't necessarily a _bad_ thing), but it's going to cause a lot of confusion.

For now, we're going to leave `rules_android` pinned to an older version to avoid this. 

I've not spent a huge amount of time looking into this repo, but why does it need to contribute to the root project's namespace?

**Comment by shs96c** [comment]
FWIW, I've also worked at companies which require every dependency (transitive or not) being used in production to be checked by their legal departments for compliance with license agreements. Having no way to prevent dependencies being added to the main namespace makes that job more difficult.

**Comment by agluszak** [comment]
So this issue seems to have a pretty wide impact: in rules_kotlin we cannot bump dependencies because of it. See https://github.com/bazelbuild/rules_kotlin/pull/1400

The interesting part is that that using bazel_worker_java 0.0.6, rules_android 0.6.4 and protobuf 29.0 seemed to be fine - https://github.com/bazelbuild/rules_kotlin/pull/1400/files/ddab11bc00697524619777f8e1e9b268e9ab49cc

But after we bumped protobuf to 33.1 (https://github.com/bazelbuild/rules_kotlin/pull/1400) this started to happen:

```
ERROR: /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/external/bazel_worker_java+/src/main/java/com/google/devtools/build/lib/worker/BUILD.bazel:5:13: no such target '@@rules_jvm_external++maven+maven//:com_google_protobuf_protobuf_java_util': target 'com_google_protobuf_protobuf_java_util' not declared in package '' defined by /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/external/rules_jvm_external++maven+maven/BUILD and referenced by '@@bazel_worker_java+//src/main/java/com/google/devtools/build/lib/worker:work_request_handlers'
```
